frontend/lib/src/hooks/useOnInputChange.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Dispatch, SetStateAction, useCallback } from "react"
+
+import { isInForm } from "@streamlit/lib/src/util/utils"
+import { ValueWithSource } from "@streamlit/lib/src/hooks/useBasicWidgetState"
+
+type OnInputChangeEventType = {
+  target: { value: string }
+} & Partial<HTMLInputElement>
+
+/**
+ * Will return a memoized function that accepts an HTMLInputElement and will call
+ * commitWidgetValue and setDirty with its value, unless the value is longer than
+ * maxChars. Will also call the setValueWithSource callback if the input is in a form.
+ *
+ * @param formId if is in a form
+ * @param maxChars if the input element's value length is greater than this, nothing will be called. Set to 0 to disable.
+ * @param setDirty calls setDirty with true
+ * @param setUiValue calls setUiValue with the input element's value
+ * @param setValueWithSource calls setValueWithSource with the input element's value
+ * @return memoized callback
+ */
+export default function useOnInputChange(
+  formId: string | undefined,
+  maxChars: number,
+  setDirty: (dirty: boolean) => void,
+  setUiValue: (value: string) => void,
+  setValueWithSource: Dispatch<
+    SetStateAction<ValueWithSource<string | null> | null>
+  >
```

Comment:
**suggestion (non-blocking):** This uses 5 positional arguments, which can be challenging to manage. To enhance the developer experience, consider refactoring this to use an object instead.

Reply:
Good idea, will update it 👍